tracer/src/Datadog.Trace/Debugger/SpanCodeOrigin/SpanCodeOriginManager.cs

```diff
@@ -0,0 +1,150 @@
+{
+    internal class SpanCodeOriginManager
+    {
+        private const string CodeOriginTag = "_dd.code_origin";
+
+        private const string FramesPrefix = "frames";
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(SpanCodeOriginManager));
+
+        private static object _globalInstanceLock = new();
+
+        private static bool _globalInstanceInitialized;
+
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring as nullable.
+        private static SpanCodeOriginManager _instance;
+#pragma warning restore CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring as nullable.
+
+        private readonly DebuggerSettings _settings = LiveDebugger.Instance?.Settings ?? DebuggerSettings.FromDefaultSource();
+
+        internal static SpanCodeOriginManager Instance =>
+            LazyInitializer.EnsureInitialized(
+                ref _instance,
+                ref _globalInstanceInitialized,
+                ref _globalInstanceLock);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void SetCodeOrigin(Span? span)
+        {
+            if (span == null || !this._settings.CodeOriginForSpansEnabled)
+            {
+                return;
+            }
+
+            AddExitSpanTag(span);
+        }
+
+        private void AddExitSpanTag(Span span)
+        {
+            var frames = ArrayPool<FrameInfo>.Shared.Rent(this._settings.CodeOriginMaxUserFrames);
+            try
+            {
+                var framesLength = PopulateUserFrames(frames);
+                if (framesLength == 0)
+                {
+                    Log.Warning("No user frames were found");
+                    return;
+                }
+
+                span.Tags.SetTag($"{CodeOriginTag}.type", "exit");
+                for (var i = 0; i < framesLength; i++)
+                {
+                    ref var info = ref frames[i];
+
+                    // PopulateUserFrames returns only frames that have method
+                    var method = info.Frame.GetMethod()!;
+                    var type = method.DeclaringType?.FullName ?? method.DeclaringType?.Name;
+                    if (string.IsNullOrEmpty(type))
+                    {
+                        continue;
+                    }
+
+                    span.Tags.SetTag($"{CodeOriginTag}.{FramesPrefix}.{i}.index", info.FrameIndex.ToString());
+                    span.Tags.SetTag($"{CodeOriginTag}.{FramesPrefix}.{i}.method", method.Name);
+                    span.Tags.SetTag($"{CodeOriginTag}.{FramesPrefix}.{i}.type", type);
+
+                    var fileName = info.Frame.GetFileName();
+                    if (!string.IsNullOrEmpty(fileName))
+                    {
+                        // todo: should we normalize?
```

Comment:
Does this TODO need addressing now?

Reply:
This will be addressed in the next code origin PR.